### PR TITLE
OKTA-326876: ui testing implementation with Xamarin.UITest

### PR DESCRIPTION
**Pull request by bryanapellanes-okta**
## Issue \#
[OKTA-326876](https://oktainc.atlassian.net/browse/OKTA-326876)

## Code
This code change includes work done to convert the existing `Android` and `iOS` applications to use `Xamarin.Forms`.  However, the files relevant to the topic are these:
- [UITest-Android.sh](https://github.com/okta/okta-oidc-xamarin/pull/16/files#diff-1e01ccfd7818eaaf345d0ae8c8da40ba)
- [UITest-iOS.sh](https://github.com/okta/okta-oidc-xamarin/pull/16/files#diff-74f9faafadd4b18f0d5fafe97de0fd85)

Also, peripherally relevant are all the shell scripts prefixed with `build-` which are used for testing during the development process.

### CODE REVIEW

File: Okta.Xamarin/Okta.Xamarin.Android/ActivityCustomUrlSchemeInterceptor.cs [hunk 1]
```diff
@@ -2,20 +2,38 @@
...
+			actions: new[] { Intent.ActionView }, 
+			Categories = new[] { Intent.CategoryDefault, Intent.CategoryBrowsable },
+			DataSchemes = new[] { "com.okta.xamarin.android" },
+			DataPath = "/callback"
+		)
+    ]
 	public class ActivityCustomUrlSchemeInterceptor : Activity
 	{
-		protected override void OnCreate(Bundle savedInstanceState)
+		protected override async void OnCreate(Bundle savedInstanceState)
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.Android/ActivityCustomUrlSchemeInterceptor.cs:36]: Should this method use the `Async` suffix since we have changed it to async?

File: Okta.Xamarin/Okta.Xamarin.Android/Assets/AboutAssets.txt [hunk 2]
```diff
@@ -0,0 +1,19 @@
+Any raw assets you want to be deployed with your application can be placed in
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.Android/Assets/AboutAssets.txt:1]: Is this file necessary?

File: Okta.Xamarin/Okta.Xamarin.Android/MainActivity.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+using Android.App;
+using Android.Content.PM;
+using Android.Runtime;
+using Android.Views;
+using Android.Widget;
+using Android.OS;
+using Android.Content;
+using Xamarin.Forms;
+
+namespace Okta.Xamarin.Droid
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.Android/MainActivity.cs:12]: Is this the expected namespace? Should it be `Okta.Xamarin.Android` instead?

File: Okta.Xamarin/Okta.Xamarin.Android/MainActivity.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        protected override void OnCreate(Bundle savedInstanceState)
+        {
+			OktaContext.Init(new OidcClient(this, OktaConfig.LoadFromXmlStream(Assets.Open("OktaConfig.xml"))));
+			OktaContext.Current.SignInCompleted += (sender, args) => Shell.Current.GoToAsync("//ProfilePage");
+
+			TabLayoutResource = Resource.Layout.Tabbar;
+            ToolbarResource = Resource.Layout.Toolbar;
+
+            base.OnCreate(savedInstanceState);
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.Android/MainActivity.cs:25]: Nit-pick: We should fix the indentation here.

File: Okta.Xamarin/Okta.Xamarin.Android/Okta.Xamarin.Android.csproj [hunk 5]
```diff
@@ -54,39 +89,65 @@
...
     <None Include="Resources\AboutResources.txt" />
+    <None Include="Assets\AboutAssets.txt" />
+    <None Include="Properties\AndroidManifest.xml" />
   </ItemGroup>
   <ItemGroup>
-    <PackageReference Include="Xamarin.Android.Support.CustomTabs">
-      <Version>28.0.0.1</Version>
-    </PackageReference>
-    <PackageReference Include="Xamarin.Android.Support.v7.AppCompat" Version="28.0.0.1" />
+    <AndroidResource Include="Resources\layout\Tabbar.xml" />
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.Android/Okta.Xamarin.Android.csproj:116]: Since we do not have any native view yet, should we exclude these resources for now? Do we need them?

File: Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Okta.Xamarin.UITest
+{
+	// TODO: move OktaConfig to Okta.Xamarin.Shared to eliminate the need for this class
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs:9]: Is this TODO still valid?

File: Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+		public TimeSpan ClockSkew { get; set; }
+
+		// TODO: refactor IOktaConfig and related to consolidate implementation
+		/// <summary>
+		/// Gets the Authorize Url used for logging in, which is either the <see cref="AuthorizeUri"/> if specified, or constructed from the <see cref="OktaDomain"/> and <see cref="AuthorizationServerId"/>.
+		/// </summary>
+		/// <returns>The computed Authorize Url used for logging in</returns>
+		public string GetAuthorizeUrl()
+		{
+			if (string.IsNullOrEmpty(AuthorizeUri))
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs:34]: Nit-pick: Fix indentation.

File: Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
+﻿using System;
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs:1]: This can be probably managed in a diff PR, but all files should include the Okta header.

File: Okta.Xamarin/Okta.Xamarin.UITest.Shared/UITestData.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+using Okta.Xamarin;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Net.NetworkInformation;
+
+namespace Okta.Xamarin.UITest
+{
+	public class UITestData
+	{
+		public static UITestData Load(string filePath = "~/.okta/UITestData.json")
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.UITest.Shared/UITestData.cs:16]: Nit-pick: The indentation here looks different from other files. I think it'd be a good idea to use a linter (Stylecop for example) for consistency.

File: Okta.Xamarin/Okta.Xamarin/IOidcClient.cs [hunk 10]
```diff
@@ -12,10 +12,11 @@ namespace Okta.Xamarin
 	/// </summary>
 	public interface IOidcClient
 	{
+		object ReturnContext { get; set; }
 		/// <summary>
-		/// The configuration for this Client.  Must be set in the constructor.
+		/// The configuration for this Client.
 		/// </summary>
-		IOktaConfig Config { get; }
+		IOktaConfig Config { get; set; }
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/IOidcClient.cs:19]: Shouldn't the config be immutable?

File: Okta.Xamarin/Okta.Xamarin/Models/Item.cs [hunk 11]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Okta.Xamarin.Models
+{
+    public class Item
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/Models/Item.cs:5]: Do we need this class?

File: Okta.Xamarin/Okta.Xamarin/OidcClient.cs [hunk 12]
```diff
@@ -19,9 +19,9 @@ namespace Okta.Xamarin
 	public partial class OidcClient : IOidcClient
 	{
 		/// <summary>
-		/// The configuration for this Client.  Must be set in the constructor.
+		/// The configuration for this Client.
 		/// </summary>
-		public IOktaConfig Config { get; private set; }
+		public IOktaConfig Config { get; set; }
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/OidcClient.cs:24]: I'm curious, why are we changing this? I believe the configuration should remain immutable. Once you create a client with a specific config, how is this gonna work if I change the config afterward?

File: Okta.Xamarin/Okta.Xamarin/OktaContext.cs [hunk 13]
```diff
@@ -0,0 +1,61 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Okta.Xamarin
+{
+	public class OktaContext<Cl, Co> : OktaContext
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/OktaContext.cs:8]: Is it possible to use more meaningful names here? Client and Config instead of Cl and Co?

File: Okta.Xamarin/Okta.Xamarin/Services/IDataStore.cs [hunk 14]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Okta.Xamarin.Services
+{
+    public interface IDataStore<T>
+    {
+        Task<bool> AddItemAsync(T item);
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/Services/IDataStore.cs:9]: We document public methods.

File: Okta.Xamarin/Okta.Xamarin/Services/MockDataStore.cs [hunk 15]
```diff
@@ -0,0 +1,60 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Okta.Xamarin.Models;
+
+namespace Okta.Xamarin.Services
+{
+    public class MockDataStore : IDataStore<Item>
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/Services/MockDataStore.cs:9]: Is there any reason for having testing models in the Xamarin project? Are you planning to move these to a separate test project in the future?

File: Okta.Xamarin/Okta.Xamarin/Views/NewItemPage.xaml.cs [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+using System.ComponentModel;
+using Xamarin.Forms;
+using Xamarin.Forms.Xaml;
+
+using Okta.Xamarin.Models;
+using Okta.Xamarin.ViewModels;
+
+namespace Okta.Xamarin.Views
+{
+    public partial class NewItemPage : ContentPage
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/Views/NewItemPage.xaml.cs:12]: Do we need this class?

File: build.cake [hunk 17]
```diff
@@ -1,3 +1,9 @@
+#addin "nuget:?package=Cake.Xamarin&version=3.1.0"
+
+Action<string> write = (msg) =>
```
Comment (laura-rodriguez) [inline build.cake:3]: Nit-pick: rename write to Write for consistency.

File: build.cake [hunk 18]
```diff
@@ -168,13 +190,21 @@ Task("Pack-Android")
...
+    });
+
 Task("Build-iOS")
     .IsDependentOn("Restore-Packages")
     .Does (() =>
     {
         MSBuild(iOSProject, settings => 
             settings
-                .SetConfiguration(configuration)   
+                .SetConfiguration("Debug")
```
Comment (laura-rodriguez) [inline build.cake:207]: Did you mean to commit this change? What happens when you want to pack with config = release?

File: commit [hunk 19]
```diff
@@ -1 +1 @@
-099dee4
+875702e
```
Comment (laura-rodriguez) [inline commit:1]: What is this? Did you mean to include it?

File: nuget/semver/major [hunk 20]
```diff
@@ -0,0 +1 @@
+0
```
Comment (laura-rodriguez) [inline nuget/semver/major:1]: what is this? did you mean to include it?

File: tickle [hunk 21]
```diff
@@ -1 +1 @@
-1
+Fri, Sep 25, 2020 7:19:00 AM
```
Comment (laura-rodriguez) [inline tickle:1]: Did you mean to include this?

File: Okta.Xamarin/Okta.Xamarin.Android/ActivityCustomUrlSchemeInterceptor.cs [hunk 22]
```diff
@@ -2,20 +2,38 @@
...
+using Okta.Xamarin.Android;
 
 namespace Okta.Xamarin.Android
 {
 	/// <summary>
 	/// This is used to handle the callback from the Chrome custom tab after a user logs in.  You need to create a new activity that inherits from this class and set the appropriate IntentFilter.  Please see the documentation for more information: https://github.com/okta/okta-oidc-xamarin/blob/master/README.md#registering-callbacks
 	/// </summary>
+	[Activity(Label = "ActivityCustomUrlSchemeInterceptor", NoHistory = true, LaunchMode = LaunchMode.SingleTop)]
+	[
+		IntentFilter
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin.Android/ActivityCustomUrlSchemeInterceptor.cs:26]: This file is using tabs instead of spaces

**Review by laura-rodriguez — COMMENTED** [review]
Finished my first pass. Please review my comments.

I believe that a linter will help with indentation issues and missing documentation.

File: Okta.Xamarin/Okta.Xamarin.Android/ActivityCustomUrlSchemeInterceptor.cs (hunk 1 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.Android/ActivityCustomUrlSchemeInterceptor.cs:36]: This method overrides a method defined on the `Activity` class, renaming it would invalidate the override.  Declaring it `async void` allows the implementation to `await` if necessary.

File: Okta.Xamarin/Okta.Xamarin.Android/Assets/AboutAssets.txt (hunk 2 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.Android/Assets/AboutAssets.txt:1]: This file was added by the VS Xamarin forms project template.  Since I intend to use this project as the basis for a future VS template which extends the base template, it should be left in place so our version of the template can be treated as an extension to the base template.

File: Okta.Xamarin/Okta.Xamarin.Android/MainActivity.cs (hunk 3 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.Android/MainActivity.cs:12]: This namespace is part of the VS Xamarin project template and is used extensively by generated code; attempting to change it has caused a number of issues.

File: Okta.Xamarin/Okta.Xamarin.Android/Okta.Xamarin.Android.csproj (hunk 5 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.Android/Okta.Xamarin.Android.csproj:116]: These files are used by the Xamarin forms shell and would cause issues if removed.

File: Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs (hunk 6 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs:9]: Yes.  There are a number of redundant classes that will add to confusion in the future if not consolidated; this class is one.

File: Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs (hunk 7 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs:34]: ❓ I don't see indentation issue here; I'll submit future pull request with style cop formatting changes.

File: Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs (hunk 8 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs:1]: I'll add follow up item to address this along with formatting.

File: Okta.Xamarin/Okta.Xamarin/IOidcClient.cs (hunk 10 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/IOidcClient.cs:19]: Order of operations issues and a need to be able to construct a client without parameters for both Android and iOS necessitated this setter be made public.  I can plan to make it "lazy immutable" by ensuring it is set only once where appropriate.

File: Okta.Xamarin/Okta.Xamarin/OidcClient.cs (hunk 12 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/OidcClient.cs:24]: See previous comment (https://github.com/okta/okta-oidc-xamarin/pull/16#discussion_r496137768).  To address your concern I can plan to make this settable only once, however, resetting the config won't necessarily have any adverse affects.

File: Okta.Xamarin/Okta.Xamarin/Services/IDataStore.cs (hunk 14 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/Services/IDataStore.cs:9]: This file is part of the Xamarin Forms template project and is used by the base application shell.  It is expected that this file will be modified or removed for the purposes of the consumer of the project.

File: Okta.Xamarin/Okta.Xamarin/Services/MockDataStore.cs (hunk 15 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/Services/MockDataStore.cs:9]: There is additional refactoring that needs to be done to better organize the project structure for future extensibility; this file along with others should be moved as you suggest (looking closer, this comment also applies here: https://github.com/okta/okta-oidc-xamarin/pull/16#discussion_r496144379)

File: Okta.Xamarin/Okta.Xamarin/Views/NewItemPage.xaml.cs (hunk 16 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/Views/NewItemPage.xaml.cs:12]: This class is used by the Xamarin forms shell.  We _may_ choose to refactor it away in the future.

File: build.cake (hunk 17 above)
Comment (bryanapellanes-okta) [reply build.cake:3]: Not sure I agree that capitalizing will add to consistency.  `write` is a local variable in this context, similarly named to the other functions below in the same file, for example: https://github.com/okta/okta-oidc-xamarin/pull/16/files/3a43502b3a1f2f66fb55f1ff7c4a8bda21fd4576#diff-4d3634747586b964df9f7712bc3ec669R96

File: build.cake (hunk 18 above)
Comment (bryanapellanes-okta) [reply build.cake:207]: Should've added that instead of changing.  I'll update.

File: commit (hunk 19 above)
Comment (bryanapellanes-okta) [reply commit:1]: This file is part of release automation scripts and should be ignored by humans.

File: nuget/semver/major (hunk 20 above)
Comment (bryanapellanes-okta) [reply nuget/semver/major:1]: Files in this folder are part of release automation and handle versioning.

File: tickle (hunk 21 above)
Comment (bryanapellanes-okta) [reply tickle:1]: If a previous build is not available for the current state of the code, this file is used for release build automation to start a new build without making any functional changes.  This is useful in cases where the previous build was deleted.

File: Okta.Xamarin/Okta.Xamarin/Models/Item.cs (hunk 11 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/Models/Item.cs:5]: This class is used by the Xamarin Forms views for the default shell/UI.

**Comment by bryanapellanes-okta** [comment]
@laura-rodriguez @andriizhegurov-okta I changed the target branch from master to `version-0.0.1`.  I'll plan to address formatting in separate PR targeting the same branch.

**Review by andriizhegurov-okta — APPROVED** [review]
👍
